src/particle.cpp

```diff
@@ -119,6 +132,506 @@ Particle::from_source(const Bank* src)
+    }
+
+
+    if (!model::active_meshsurf_tallies.empty()) {
+      Position r {coord[0].xyz};
+      coord[0].xyz[0] -= TINY_BIT * coord[0].uvw[0];
+      coord[0].xyz[1] -= TINY_BIT * coord[0].uvw[1];
+      coord[0].xyz[2] -= TINY_BIT * coord[0].uvw[2];
+      score_surface_tally(this, model::active_meshsurf_tallies);
+      std::copy(&r.x, &r.x + 3, coord[0].xyz);
+    }
+
+    // Reflect particle off surface
+    Direction u = surf->reflect(coord[0].xyz, coord[0].uvw);
+
+    // Make sure new particle direction is normalized
+    double norm = u.norm();
+    coord[0].uvw[0] = u.x/norm;
+    coord[0].uvw[1] = u.y/norm;
+    coord[0].uvw[2] = u.z/norm;
+
+    // Reassign particle's cell and surface
+    coord[0].cell = last_cell[last_n_coord - 1];
+    surface = -surface;
+
+    // If a reflective surface is coincident with a lattice or universe
+    // boundary, it is necessary to redetermine the particle's coordinates in
+    // the lower universes.
+
+    n_coord = 1;
+    if (!find_cell(this, true)) {
+      this->mark_as_lost("Couldn't find particle after reflecting from surface "
+        + std::to_string(surf->id_) + ".");
+      return;
+    }
+
+    // Set previous coordinate going slightly past surface crossing
+    last_xyz_current[0] = coord[0].xyz[0] + TINY_BIT*coord[0].uvw[0];
+    last_xyz_current[1] = coord[0].xyz[1] + TINY_BIT*coord[0].uvw[1];
+    last_xyz_current[2] = coord[0].xyz[2] + TINY_BIT*coord[0].uvw[2];
+
+    // Diagnostic message
+    if (settings::verbosity >= 10 || simulation::trace) {
+      write_message("    Reflected from surface " + std::to_string(surf->id_));
+    }
+    return;
+
+  } else if (surf->bc_ == BC_PERIODIC && settings::run_mode != RUN_MODE_PLOTTING) {
+    // =======================================================================
+    // PERIODIC BOUNDARY
+
+    // Do not handle periodic boundary conditions on lower universes
+    if (n_coord != 1) {
+      this->mark_as_lost("Cannot transfer particle " + std::to_string(id) +
+        " across surface in a lower universe. Boundary conditions must be "
+        "applied to root universe.");
+      return;
+    }
+
+    // Score surface currents since reflection causes the direction of the
+    // particle to change -- artificially move the particle slightly back in
+    // case the surface crossing is coincident with a mesh boundary
+    if (!model::active_meshsurf_tallies.empty()) {
+      Position r {coord[0].xyz};
+      coord[0].xyz[0] -= TINY_BIT * coord[0].uvw[0];
+      coord[0].xyz[1] -= TINY_BIT * coord[0].uvw[1];
+      coord[0].xyz[2] -= TINY_BIT * coord[0].uvw[2];
+      score_surface_tally(this, model::active_meshsurf_tallies);
+      std::copy(&r.x, &r.x + 3, coord[0].xyz);
+    }
+
+    // Get a pointer to the partner periodic surface.  Offset the index to
+    // correct for C vs. Fortran indexing.
+    auto surf_p = dynamic_cast<PeriodicSurface*>(surf);
+    if (surf_p) {
```

Comment:
Why the `if (surf_p)` check?  It seems if the surface is not a `PeriodicSurface` then something has gone very wrong and this `if` will be no help anyway.

Reply:
Fair point, I'll remove the check